File: drivers/dart/test/value_codec_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:reddb/reddb.dart';
import 'package:reddb/src/redwire/value_codec.dart';
import 'package:test/test.dart';

void main() {
  group('ValueCodec', () {
    test('value tag table is pinned', () {
      expect(ValueCodec.tagNull, 0x00);
      expect(ValueCodec.tagBool, 0x01);
      expect(ValueCodec.tagInt, 0x02);
      expect(ValueCodec.tagFloat, 0x03);
      expect(ValueCodec.tagText, 0x04);
      expect(ValueCodec.tagBytes, 0x05);
      expect(ValueCodec.tagVector, 0x06);
      expect(ValueCodec.tagJson, 0x07);
      expect(ValueCodec.tagTimestamp, 0x08);
      expect(ValueCodec.tagUuid, 0x09);
    });

    test('encodes scalar values', () {
      expect(ValueCodec.encodeValue(null), equals(<int>[0x00]));
      expect(ValueCodec.encodeValue(true), equals(<int>[0x01, 0x01]));
      expect(ValueCodec.encodeValue(false), equals(<int>[0x01, 0x00]));
      expect(
        ValueCodec.encodeValue(1),
        equals(<int>[0x02, 0x01, 0, 0, 0, 0, 0, 0, 0]),
      );
      expect(
        ValueCodec.encodeValue(-1),
        equals(<int>[0x02, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff]),
      );
      expect(
        ValueCodec.encodeValue('x'),
        equals(<int>[0x04, 0x01, 0, 0, 0, 0x78]),
      );
    });

    test('encodes bytes timestamp uuid and json', () {
      expect(
        ValueCodec.encodeValue(Uint8List.fromList([0xde, 0xad, 0xbe, 0xef])),
        equals(<int>[0x05, 0x04, 0, 0, 0, 0xde, 0xad, 0xbe, 0xef]),
      );

      final ts = ValueCodec.encodeValue(
        DateTime.fromMillisecondsSinceEpoch(1700000000123, isUtc: true),
      );
      expect(ts[0], ValueCodec.tagTimestamp);
      expect(
        ByteData.sublistView(ts, 1).getInt64(0, Endian.little),
        1700000000,
      );

      final uuid = ValueCodec.encodeValue(
        Value.uuid('00112233-4455-6677-8899-aabbccddeeff'),
      );
      expect(uuid[0], ValueCodec.tagUuid);
      expect(
        Uint8List.sublistView(uuid, 1),
        equals(<int>[
          0x00,
          0x11,
          0x22,
          0x33,
          0x44,
          0x55,
          0x66,
          0x77,
          0x88,
          0x99,
          0xaa,
          0xbb,
          0xcc,
          0xdd,
          0xee,
          0xff,
        ]),
      );

      final json = ValueCodec.encodeValue(Value.json({'b': 2, 'a': 1}));
      expect(json[0], ValueCodec.tagJson);
      expect(ByteData.sublistView(json, 1).getUint32(0, Endian.little), 13);
      expect(utf8.decode(Uint8List.sublistView(json, 5)), '{"a":1,"b":2}');
    });

    test('encodes vector from Float32List and List<double>', () {
      final encoded = ValueCodec.encodeValue(
        Float32List.fromList([1.0, 2.0, -0.5]),
      );
      expect(encoded[0], ValueCodec.tagVector);
      expect(ByteData.sublistView(encoded, 1).getUint32(0, Endian.little), 3);
      expect(
        ByteData.sublistView(encoded, 5).getFloat32(0, Endian.little),
        1.0,
      );
      expect(
        ByteData.sublistView(encoded, 9).getFloat32(0, Endian.little),
        2.0,
      );
      expect(
        ByteData.sublistView(encoded, 13).getFloat32(0, Endian.little),
        -0.5,
      );

      final fromList = ValueCodec.encodeValue(<double>[1.5, -2.5]);
      expect(fromList[0], ValueCodec.tagVector);
      expect(ByteData.sublistView(fromList, 1).getUint32(0, Endian.little), 2);
    });

    test('encodes query with params payload', () {
      final encoded = ValueCodec.encodeQueryWithParams('Q', [42, 'x', null]);
      expect(ByteData.sublistView(encoded, 0).getUint32(0, Endian.little), 1);
      expect(encoded[4], 'Q'.codeUnitAt(0));
      expect(ByteData.sublistView(encoded, 5).getUint32(0, Endian.little), 3);
      expect(encoded[9], ValueCodec.tagInt);
      expect(ByteData.sublistView(encoded, 10).getInt64(0, Endian.little), 42);
      expect(encoded[18], ValueCodec.tagText);
      expect(
        Uint8List.sublistView(encoded, 19, 24),
        equals(<int>[1, 0, 0, 0, 0x78]),
      );
      expect(encoded[24], ValueCodec.tagNull);
      expect(encoded.length, 25);
    });

    test('http params use JSON envelopes for tagged values', () {
      final params = ValueCodec.toHttpParams([
        null,
        true,
        42,
        1.5,
        'txt',
        Uint8List.fromList(utf8.encode('hi')),
        Float32List.fromList([1.0, 2.0]),
        Value.json({'b': 2, 'a': 1}),
        DateTime.fromMillisecondsSinceEpoch(1700000000123, isUtc: true),
        Value.uuid('00112233-4455-6677-8899-aabbccddeeff'),
      ]);

      expect(params, [
        null,
        true,
        42,
        1.5,
        'txt',
        {'\$bytes': 'aGk='},
        [1.0, 2.0],
        {'a': 1, 'b': 2},
        {'\$ts': 1700000000},
        {'\$uuid': '00112233-4455-6677-8899-aabbccddeeff'},
      ]);
    });

    test('shared parameter fixtures match manifest', () {
      final manifest = jsonDecode(
        File('../../testdata/conformance/redwire/params/manifest.json')
            .readAsStringSync(),
      ) as Map<String, dynamic>;

      for (final fixture in manifest['values'] as List<dynamic>) {
        final item = fixture as Map<String, dynamic>;
        expect(
          _hex(ValueCodec.encodeValue(_fixtureValue(item['name'] as String))),
          item['redwire_hex'],
          reason: item['name'] as String,
        );
      }

      final query = (manifest['queries'] as List<dynamic>).first as Map<String, dynamic>;
      final params = [
        for (final name in query['params'] as List<dynamic>) _fixtureValue(name as String),
      ];
      expect(
        _hex(ValueCodec.encodeQueryWithParams(query['sql'] as String, params)),
        query['redwire_hex'],
        reason: query['name'] as String,
      );
    });
  });
}

Object? _fixtureValue(String name) {
  switch (name) {
    case 'null':
      return null;
    case 'bool_true':
      return true;
    case 'bool_false':
      return false;
    case 'int_min':
      return -0x8000000000000000;
    case 'int_max':
      return 0x7fffffffffffffff;
    case 'int_42':
      return 42;
    case 'float_nan':
      return _doubleFromBits(0x7ff8000000000000);
    case 'float_pos_inf':
      return double.infinity;
    case 'float_neg_inf':
      return double.negativeInfinity;
    case 'float_subnormal_min':
      return _doubleFromBits(1);
    case 'text_unicode':
      return 'h\u00e9llo';
    case 'text_x':
      return 'x';
    case 'bytes_empty':
      return Uint8List(0);
    case 'bytes_deadbeef':
      return Uint8List.fromList([0xde, 0xad, 0xbe, 0xef]);
    case 'bytes_256':
      return Uint8List.fromList(List<int>.generate(256, (i) => i));
    case 'json_nested':
      return Value.json({
        'z': [1, {'deep': [true, false]}],
        'a': null,
      });
    case 'timestamp_zero':
      return Value.timestamp(0);
    case 'timestamp_max':
      return Value.timestamp(0x7fffffffffffffff);
    case 'uuid_001122':
      return Value.uuid('00112233-4455-6677-8899-aabbccddeeff');
    case 'vector_empty':
      return Float32List(0);
    case 'vector_three':
      return Float32List.fromList([1.0, 2.0, -0.5]);
    case 'vector_128':
      return Float32List.fromList(
        List<double>.generate(128, (i) => i.toDouble()),
      );
    default:
      throw ArgumentError('unknown fixture $name');
  }
}

double _doubleFromBits(int bits) {
  final bytes = ByteData(8)..setUint64(0, bits, Endian.host);
  return bytes.getFloat64(0, Endian.host);
}

String _hex(Uint8List bytes) =>
    bytes.map((byte) => byte.toRadixString(16).padLeft(2, '0')).join();

File: drivers/zig/tests/value_codec_test.zig
const std = @import("std");
const reddb = @import("reddb");
const value_codec = reddb.redwire.value_codec;

const t = std.testing;
const empty_vector = [_]f32{};
const vector_three = [_]f32{ 1.0, 2.0, -0.5 };
const bytes_256 = makeBytes256();
const vector_128 = makeVector128();

fn makeBytes256() [256]u8 {
    var out: [256]u8 = undefined;
    for (&out, 0..) |*byte, i| {
        byte.* = @intCast(i);
    }
    return out;
}

fn makeVector128() [128]f32 {
    var out: [128]f32 = undefined;
    for (&out, 0..) |*value, i| {
        value.* = @floatFromInt(i);
    }
    return out;
}

fn readFixtureManifest(allocator: std.mem.Allocator) ![]u8 {
    const candidates = [_][]const u8{
        "../../crates/reddb-wire/tests/fixtures/params/manifest.json",
        "crates/reddb-wire/tests/fixtures/params/manifest.json",
        "../../../crates/reddb-wire/tests/fixtures/params/manifest.json",
    };

    const io = std.Options.debug_io;
    for (candidates) |path| {
        const buffer = try allocator.alloc(u8, 1024 * 1024);
        errdefer allocator.free(buffer);
        const bytes = std.Io.Dir.cwd().readFile(io, path, buffer) catch |err| switch (err) {
            error.FileNotFound => {
                allocator.free(buffer);
                continue;
            },
            else => return err,
        };
        if (bytes.len == buffer.len) return buffer;
        return try allocator.realloc(buffer, bytes.len);
    }
    return error.FixtureManifestNotFound;
}

fn manifestHex(manifest: std.json.Value, section: []const u8, name: []const u8) ![]const u8 {
    const items = manifest.object.get(section).?.array.items;
    for (items) |item| {
        const object = item.object;
        if (std.mem.eql(u8, object.get("name").?.string, name)) {
            return object.get("redwire_hex").?.string;
        }
    }
    return error.FixtureNotFound;
}

fn hexToBytes(allocator: std.mem.Allocator, hex: []const u8) ![]u8 {
    if (hex.len % 2 != 0) return error.InvalidHex;
    const out = try allocator.alloc(u8, hex.len / 2);
    errdefer allocator.free(out);
    for (out, 0..) |*byte, i| {
        byte.* = try std.fmt.parseInt(u8, hex[i * 2 .. i * 2 + 2], 16);
    }
    return out;
}

fn f64FromBits(bits: u64) f64 {
    return @bitCast(bits);
}

fn fixtureValue(name: []const u8) !value_codec.Value {
    if (std.mem.eql(u8, name, "null")) return .{ .null = {} };
    if (std.mem.eql(u8, name, "bool_true")) return .{ .bool = true };
    if (std.mem.eql(u8, name, "bool_false")) return .{ .bool = false };
    if (std.mem.eql(u8, name, "int_min")) return .{ .int = std.math.minInt(i64) };
    if (std.mem.eql(u8, name, "int_max")) return .{ .int = std.math.maxInt(i64) };
    if (std.mem.eql(u8, name, "int_42")) return .{ .int = 42 };
    if (std.mem.eql(u8, name, "float_nan")) return .{ .float = f64FromBits(0x7ff8000000000000) };
    if (std.mem.eql(u8, name, "float_pos_inf")) return .{ .float = std.math.inf(f64) };
    if (std.mem.eql(u8, name, "float_neg_inf")) return .{ .float = -std.math.inf(f64) };
    if (std.mem.eql(u8, name, "float_subnormal_min")) return .{ .float = f64FromBits(1) };
    if (std.mem.eql(u8, name, "text_unicode")) return .{ .text = "h\xc3\xa9llo" };
    if (std.mem.eql(u8, name, "text_x")) return .{ .text = "x" };
    if (std.mem.eql(u8, name, "bytes_empty")) return .{ .bytes = &.{} };
    if (std.mem.eql(u8, name, "bytes_deadbeef")) return .{ .bytes = &.{ 0xde, 0xad, 0xbe, 0xef } };
    if (std.mem.eql(u8, name, "bytes_256")) return .{ .bytes = &bytes_256 };
    if (std.mem.eql(u8, name, "json_nested")) return .{ .json = "{\"a\":null,\"z\":[1,{\"deep\":[true,false]}]}" };
    if (std.mem.eql(u8, name, "timestamp_zero")) return .{ .timestamp = 0 };
    if (std.mem.eql(u8, name, "timestamp_max")) return .{ .timestamp = std.math.maxInt(i64) };
    if (std.mem.eql(u8, name, "uuid_001122")) return try value_codec.Value.uuidFromString("00112233-4455-6677-8899-aabbccddeeff");
    if (std.mem.eql(u8, name, "vector_empty")) return .{ .vector = &empty_vector };
    if (std.mem.eql(u8, name, "vector_three")) return .{ .vector = &vector_three };
    if (std.mem.eql(u8, name, "vector_128")) return .{ .vector = &vector_128 };
    return error.UnknownFixture;
}

test "value tag table is pinned" {
    try t.expectEqual(@as(u8, 0x00), @intFromEnum(value_codec.ValueTag.null));
    try t.expectEqual(@as(u8, 0x01), @intFromEnum(value_codec.ValueTag.bool));
    try t.expectEqual(@as(u8, 0x02), @intFromEnum(value_codec.ValueTag.int));
    try t.expectEqual(@as(u8, 0x03), @intFromEnum(value_codec.ValueTag.float));
    try t.expectEqual(@as(u8, 0x04), @intFromEnum(value_codec.ValueTag.text));
    try t.expectEqual(@as(u8, 0x05), @intFromEnum(value_codec.ValueTag.bytes));
    try t.expectEqual(@as(u8, 0x06), @intFromEnum(value_codec.ValueTag.vector));
    try t.expectEqual(@as(u8, 0x07), @intFromEnum(value_codec.ValueTag.json));
    try t.expectEqual(@as(u8, 0x08), @intFromEnum(value_codec.ValueTag.timestamp));
    try t.expectEqual(@as(u8, 0x09), @intFromEnum(value_codec.ValueTag.uuid));
}

test "encodes scalar values" {
    const null_value = try value_codec.encodeValue(t.allocator, .{ .null = {} });
    defer t.allocator.free(null_value);
    try t.expectEqualSlices(u8, &.{0x00}, null_value);

    const true_value = try value_codec.encodeValue(t.allocator, .{ .bool = true });
    defer t.allocator.free(true_value);
    try t.expectEqualSlices(u8, &.{ 0x01, 0x01 }, true_value);

    const int_value = try value_codec.encodeValue(t.allocator, .{ .int = -1 });
    defer t.allocator.free(int_value);
    try t.expectEqualSlices(u8, &.{
        0x02, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
    }, int_value);

    const text_value = try value_codec.encodeValue(t.allocator, .{ .text = "x" });
    defer t.allocator.free(text_value);
    try t.expectEqualSlices(u8, &.{ 0x04, 1, 0, 0, 0, 'x' }, text_value);
}

test "encodes bytes timestamp uuid json and vector" {
    const bytes_value = try value_codec.encodeValue(t.allocator, .{ .bytes = &.{ 0xde, 0xad, 0xbe, 0xef } });
    defer t.allocator.free(bytes_value);
    try t.expectEqualSlices(u8, &.{ 0x05, 4, 0, 0, 0, 0xde, 0xad, 0xbe, 0xef }, bytes_value);

    const timestamp_value = try value_codec.encodeValue(t.allocator, .{ .timestamp = 1_700_000_000 });
    defer t.allocator.free(timestamp_value);
    try t.expectEqualSlices(u8, &.{ 0x08, 0x00, 0xf1, 0x53, 0x65, 0x00, 0x00, 0x00, 0x00 }, timestamp_value);

    const uuid_value = try value_codec.encodeValue(
        t.allocator,
        try value_codec.Value.uuidFromString("00112233-4455-6677-8899-aabbccddeeff"),
    );
    defer t.allocator.free(uuid_value);
    try t.expectEqualSlices(u8, &.{
        0x09, 0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77,
        0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff,
    }, uuid_value);

    const json_value = try value_codec.encodeValue(t.allocator, .{ .json = "{\"a\":1,\"b\":2}" });
    defer t.allocator.free(json_value);
    try t.expectEqualSlices(u8, &.{
        0x07, 13, 0, 0, 0, '{', '"', 'a', '"', ':', '1', ',', '"', 'b', '"', ':', '2', '}',
    }, json_value);

    const vector = [_]f32{ 1.0, 2.0, -0.5 };
    const vector_value = try value_codec.encodeValue(t.allocator, .{ .vector = &vector });
    defer t.allocator.free(vector_value);
    try t.expectEqualSlices(u8, &.{
        0x06, 3,    0,    0,    0,
        0x00, 0x00, 0x80, 0x3f, 0x00,
        0x00, 0x00, 0x40, 0x00, 0x00,
        0x00, 0xbf,
    }, vector_value);
}

test "encodes query with params payload" {
    const params = [_]value_codec.Value{
        .{ .int = 42 },
        .{ .text = "x" },
        .{ .null = {} },
    };
    const encoded = try value_codec.encodeQueryWithParams(t.allocator, "Q", &params);
    defer t.allocator.free(encoded);
    try t.expectEqualSlices(u8, &.{
        1,  0, 0, 0,    'Q',
        3,  0, 0, 0,    0x02,
        42, 0, 0, 0,    0,
        0,  0, 0, 0x04, 1,
        0,  0, 0, 'x',  0x00,
    }, encoded);
}

test "http query body adds typed params only when non-empty" {
    const empty = try value_codec.toHttpQueryBody(t.allocator, "SELECT \"x\"", &.{});
    defer t.allocator.free(empty);
    try t.expectEqualStrings("{\"query\":\"SELECT \\\"x\\\"\"}", empty);

    const bytes = [_]u8{ 'h', 'i' };
    const vector = [_]f32{ 1.0, 2.0 };
    const params = [_]value_codec.Value{
        .{ .null = {} },
        .{ .bool = true },
        .{ .int = 42 },
        .{ .float = 1.5 },
        .{ .text = "txt" },
        .{ .bytes = &bytes },
        .{ .vector = &vector },
        .{ .json = "{\"a\":1,\"b\":2}" },
        .{ .timestamp = 1_700_000_000 },
        try value_codec.Value.uuidFromString("00112233-4455-6677-8899-aabbccddeeff"),
    };
    const body = try value_codec.toHttpQueryBody(t.allocator, "SELECT $1", &params);
    defer t.allocator.free(body);
    try t.expect(std.mem.indexOf(u8, body, "\"params\"") != null);
    try t.expect(std.mem.indexOf(u8, body, "\"$bytes\":\"aGk=\"") != null);
    try t.expect(std.mem.indexOf(u8, body, "\"$uuid\":\"00112233-4455-6677-8899-aabbccddeeff\"") != null);
}

test "shared parameter fixtures match manifest" {
    const manifest_bytes = try readFixtureManifest(t.allocator);
    defer t.allocator.free(manifest_bytes);

    var parsed = try std.json.parseFromSlice(std.json.Value, t.allocator, manifest_bytes, .{});
    defer parsed.deinit();

    const names = [_][]const u8{
        "null",
        "bool_true",
        "bool_false",
        "int_min",
        "int_max",
        "int_42",
        "float_nan",
        "float_pos_inf",
        "float_neg_inf",
        "float_subnormal_min",
        "text_unicode",
        "text_x",
        "bytes_empty",
        "bytes_deadbeef",
        "bytes_256",
        "json_nested",
        "timestamp_zero",
        "timestamp_max",
        "uuid_001122",
        "vector_empty",
        "vector_three",
        "vector_128",
    };

    for (names) |name| {
        const encoded = try value_codec.encodeValue(t.allocator, try fixtureValue(name));
        defer t.allocator.free(encoded);

        const expected = try hexToBytes(t.allocator, try manifestHex(parsed.value, "values", name));
        defer t.allocator.free(expected);

        try t.expectEqualSlices(u8, expected, encoded);
    }

    const query = parsed.value.object.get("queries").?.array.items[0].object;
    const param_names = query.get("params").?.array.items;
    const params = try t.allocator.alloc(value_codec.Value, param_names.len);
    defer t.allocator.free(params);
    for (param_names, 0..) |param, i| {
        params[i] = try fixtureValue(param.string);
    }

    const encoded_query = try value_codec.encodeQueryWithParams(
        t.allocator,
        query.get("sql").?.string,
        params,
    );
    defer t.allocator.free(encoded_query);

    const expected_query = try hexToBytes(
        t.allocator,
        try manifestHex(parsed.value, "queries", query.get("name").?.string),
    );
    defer t.allocator.free(expected_query);

    try t.expectEqualSlices(u8, expected_query, encoded_query);
}
